### PR TITLE
Support custom routes

### PR DESCRIPTION
**Pull request by xepozz**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

Let's image a request to `/user/index`.
It has possible to controller+action variations:
1. `/Controller/UserController::index()`
2. `/Controller/User/IndexController::index()`

Order to find a possible method for `/user/index`:
- UserController::index()
- User/IndexController::index()

Custom routes make it able to write routes like the following: `user/index`, `user/main`, `user/default`, etc. In general just `^user/([^\/]+)$` (read regexp as "any chars if it doesn't contain a slash after user/")

So having `GET /user/hello/world` the `FileRouter` will look for:
- `/Controller/User/Hello/WorldController::index()`, `index` is a method converted from `GET` method
- `/Controller/User/HelloController::world()`, method doesn't matter

As a confirmation in tests look at https://github.com/yiisoft/file-router/blob/5e192a6cbfcd59e8dbb2f80bd9661d5a78f9725c/tests/FileRouterTest.php#L255

### CODE REVIEW

**Review by samdark — APPROVED** [review]

File: src/FileRouter.php [hunk 1]
```diff
@@ -47,73 +47,108 @@ public function withNamespace(string $namespace): self
...
+        yield [
+            $this->cleanClassname(
+                $this->namespace . '\\' . $this->baseControllerDirectory . '\\' . $directoryPath . '\\' . $controllerName . $this->classPostfix
+            ),
+            $possibleAction,
+        ];
+    }
 
-        $className = str_replace(
+    protected function cleanClassname(string $className): string|array
```
Comment (vjik) [inline src/FileRouter.php]:
```suggestion
    private function cleanClassname(string $className): string
```

File: src/FileRouter.php [hunk 2]
```diff
@@ -47,73 +47,108 @@ public function withNamespace(string $namespace): self
...
+                $this->cleanClassname(
+                    $this->namespace . '\\' . $this->baseControllerDirectory . '\\' . $controllerName . $this->classPostfix
+                ),
+                $possibleAction,
+            ];
+            return;
+        }
+
+        $controllerName = preg_replace_callback(
+            '#(/.)#',
```
Comment (vjik) [inline src/FileRouter.php:110]: This doesn't work with unicode. For example, `привет/мир` → `привет/МирController::index`
May be extract it to separate issue?

File: src/FileRouter.php [hunk 3]
```diff
@@ -47,73 +47,108 @@ public function withNamespace(string $namespace): self
...
+            fn(array $matches) => strtoupper($matches[1]),
```
Comment (vjik) [inline src/FileRouter.php]:
```suggestion
            static fn(array $matches) => strtoupper($matches[1]),
```

File: src/FileRouter.php [hunk 4]
```diff
@@ -47,73 +47,108 @@ public function withNamespace(string $namespace): self
...
-        }
+            /** @psalm-suppress InvalidPropertyFetch */
+            $actions = $controllerClass::$actions ?? [
+                'HEAD' => 'head',
+                'OPTIONS' => 'options',
+                'GET' => 'index',
+                'POST' => 'create',
+                'PUT' => 'update',
+                'DELETE' => 'delete',
+            ];
```
Comment (vjik) [inline src/FileRouter.php:70]: Not need when `$possibleAction` is not null.

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/FileRouter.php (hunk 2 above)
Comment (xepozz) [reply src/FileRouter.php:110]: https://github.com/yiisoft/file-router/pull/8

File: src/FileRouter.php (hunk 4 above)
Comment (xepozz) [reply src/FileRouter.php:70]: https://github.com/yiisoft/file-router/pull/8/commits/abf457a5cfb1ef54f758c27a4606616a3e6c72b3#diff-0fad446e9cf870f37a58ee935b71d74a1d248c95960707b489a9e230a96d742bR63

**Review by vjik — APPROVED** [review]

**Pull request merged**